### PR TITLE
[RISCV] Introduce kernel SBI library.

### PR DESCRIPTION
**Pull request by MichalBlk**
A supervisor mode RISC-V kernel communicates with its execution environment via Supervisor Binary Interface (SBI). The presented change introduces a kernel library for issuing services defined by the SBI specification.

### CODE REVIEW

File: sys/riscv/sbi.c [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+  __asm __volatile("ecall"
+                   : "+r"(a0), "+r"(a1)
+                   : "r"(a2), "r"(a3), "r"(a4), "r"(a6), "r"(a7)
+                   : "memory");
+
+  return (sbi_ret_t){
+    .error = a0,
+    .value = a1,
+  };
+}
```
Comment (pj1031999) [inline sys/riscv/sbi.c:79]: I'm not sure if we can trust compiler enough to use that tricky syntax - from my point of view it's better to write this function in pure assembler.

File: sys/riscv/sbi.c [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+void sbi_set_timer(uint64_t val) {
+  sbi_ret_t ret;
+
+  /* Use the TIME legacy replacement extension, if available. */
+  if (has_time_extension)
+    ret = SBI_CALL1(SBI_EXT_ID_TIME, SBI_TIME_SET_TIMER, val);
+  else
+    ret = SBI_CALL1(SBI_SET_TIMER, 0, val);
+
+  assert(ret.error == SBI_SUCCESS);
```
Comment (pj1031999) [inline sys/riscv/sbi.c:133]: Why we can't propagate error and give opportunity to decide how to interpret that?

File: sys/riscv/sbi.c [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+void sbi_send_ipi(const u_long *hart_mask) {
+  sbi_ret_t ret;
+
+  /* Use the IPI legacy replacement extension, if available. */
+  if (has_ipi_extension)
+    ret = SBI_CALL2(SBI_EXT_ID_IPI, SBI_IPI_SEND_IPI, *hart_mask, 0);
+  else
+    ret = SBI_CALL1(SBI_SEND_IPI, 0, (u_long)hart_mask);
+
+  assert(ret.error == SBI_SUCCESS);
```
Comment (pj1031999) [inline sys/riscv/sbi.c:149]: ditto

File: sys/riscv/sbi.c [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+void sbi_remote_fence_i(const u_long *hart_mask) {
+  sbi_ret_t ret;
+
+  /* Use the RFENCE legacy replacement extension, if available. */
+  if (has_rfnc_extension)
+    ret = SBI_CALL2(SBI_EXT_ID_RFNC, SBI_RFNC_REMOTE_FENCE_I, *hart_mask, 0);
+  else
+    ret = SBI_CALL1(SBI_REMOTE_FENCE_I, 0, (u_long)hart_mask);
+
+  assert(ret.error == SBI_SUCCESS);
```
Comment (pj1031999) [inline sys/riscv/sbi.c:165]: ditto

File: sys/riscv/sbi.c [hunk 5]
```diff
@@ -0,0 +1,346 @@
...
+  sbi_ret_t ret;
+
+  /* Use the RFENCE legacy replacement extension, if available. */
+  if (has_rfnc_extension)
+    ret = SBI_CALL4(SBI_EXT_ID_RFNC, SBI_RFNC_REMOTE_SFENCE_VMA, *hart_mask, 0,
+                    start, size);
+  else
+    ret = SBI_CALL3(SBI_REMOTE_SFENCE_VMA, 0, (u_long)hart_mask, start, size);
+
+  assert(ret.error == SBI_SUCCESS);
```
Comment (pj1031999) [inline sys/riscv/sbi.c:178]: ditto

File: sys/riscv/sbi.c [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+
+  /* Use the RFENCE legacy replacement extension, if available. */
+  if (has_rfnc_extension)
+    ret = SBI_CALL5(SBI_EXT_ID_RFNC, SBI_RFNC_REMOTE_SFENCE_VMA_ASID,
+                    *hart_mask, 0, start, size, asid);
+  else
+    ret = SBI_CALL4(SBI_REMOTE_SFENCE_VMA_ASID, 0, (u_long)hart_mask, start,
+                    size, asid);
+
+  assert(ret.error == SBI_SUCCESS);
```
Comment (pj1031999) [inline sys/riscv/sbi.c:193]: ditto

File: sys/riscv/sbi.c [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+}
+
+/*
+ * Hart state management extension.
+ */
+
+int sbi_hsm_hart_start(u_long hart, u_long start_addr, u_long priv) {
+  sbi_ret_t ret =
+    SBI_CALL3(SBI_EXT_ID_HSM, SBI_HSM_HART_START, hart, start_addr, priv);
+  return (int)ret.error;
```
Comment (pj1031999) [inline sys/riscv/sbi.c:203]: Why result is truncated to int?

File: sys/riscv/sbi.c [hunk 8]
```diff
@@ -0,0 +1,346 @@
...
+}
+
+void sbi_hsm_hart_stop(void) {
+  (void)SBI_CALL0(SBI_EXT_ID_HSM, SBI_HSM_HART_STOP);
+}
+
+int sbi_hsm_hart_status(u_long hart) {
+  sbi_ret_t ret = SBI_CALL1(SBI_EXT_ID_HSM, SBI_HSM_HART_STATUS, hart);
+  return (ret.error != SBI_SUCCESS ? (int)ret.error : (int)ret.value);
```
Comment (pj1031999) [inline sys/riscv/sbi.c:212]: ditto

File: sys/riscv/sbi.c [hunk 9]
```diff
@@ -0,0 +1,346 @@
...
+    (void)SBI_CALL1(SBI_CONSOLE_PUTCHAR, 0, ch);
+}
+
+int sbi_console_getchar(void) {
+  if (has_console_getchar_extension) {
+    /*
+     * XXX: The "error" is returned here because legacy SBI functions
+     * continue to return their value in a0.
+     */
+    return (SBI_CALL0(SBI_CONSOLE_GETCHAR, 0).error);
```
Comment (pj1031999) [inline sys/riscv/sbi.c:243]: ditto

File: sys/riscv/sbi.c [hunk 10]
```diff
@@ -0,0 +1,346 @@
...
+static u_long sbi_spec_version;
+static u_long sbi_impl_id;
+static u_long sbi_impl_version;
+
+static bool has_time_extension = false;
+static bool has_ipi_extension = false;
+static bool has_rfnc_extension = false;
+static bool has_srst_extension = false;
+static bool has_console_putchar_extension = false;
+static bool has_console_getchar_extension = false;
```
Comment (pj1031999) [inline sys/riscv/sbi.c:53]: Not strong opinion - you can consider creating enum with or of extensions.

**Review by pj1031999 — COMMENTED** [review]
A bunch of questions. I don't have strong opinion here since this code is copied but I want to understand some quirks.

File: sys/riscv/sbi.c (hunk 1 above)
Comment (MichalBlk) [reply sys/riscv/sbi.c]: FreeBSD and Linux do it this way, however, I'm not sure whether it requires some additional compiler options to handle it properly in all scenarios (e.g. when the -O3 is used). I have changed it to relay on bare assembly.

File: sys/riscv/sbi.c (hunk 2 above)
Comment (MichalBlk) [reply sys/riscv/sbi.c]: I agree. Done.

File: sys/riscv/sbi.c (hunk 5 above)
Comment (MichalBlk) [reply sys/riscv/sbi.c]: Done.

File: sys/riscv/sbi.c (hunk 6 above)
Comment (MichalBlk) [reply sys/riscv/sbi.c]: Done.

File: sys/riscv/sbi.c (hunk 8 above)
Comment (MichalBlk) [reply sys/riscv/sbi.c:209]: as above

File: sys/riscv/sbi.c (hunk 9 above)
Comment (MichalBlk) [reply sys/riscv/sbi.c]: as above

File: sys/riscv/sbi.c (hunk 10 above)
Comment (MichalBlk) [reply sys/riscv/sbi.c]: We decided to rely on the current version of OpenSBI. The only extension whose support is platform-specific is the system reset extension, so now we have only a single flag.

Comment (MichalBlk) [reply sys/riscv/sbi.c]: The assumptions are described at the top of `sys/riscv/sbi.c`.

**Review by cahirwpz — APPROVED** [review]

**Pull request merged**